services/sidecar/rollup/derivation/batch_builder.go

```diff
@@ -43,99 +20,117 @@ type InvalidBlockError struct{ Msg string }
+			return nil, nil
+		}
+		return nil, fmt.Errorf("failed to get batch: %w", err)
 	}
-
-	b.builtBatchData = batchData
-	return b.builtBatchData, nil
+	// Cache last built batch.
+	b.lastBuilt = batch
+	return b.lastBuilt, nil
 }
 
+// Advances the builder, clearing the last built batch.
 func (b *batchBuilder) Advance() {
-	b.builtBatchData = nil
+	b.lastBuilt = nil
 }
 
-func (b *batchBuilder) serializeToAttrs() (*BatchAttributes, error) {
+// Encodes pending blocks into a new batch, constrained by `maxBatchSize`.
+// Returns an `io.EOF` error if there are no pending blocks.
+func (b *batchBuilder) encodePending() error {
 	if len(b.pendingBlocks) == 0 {
-		return nil, io.EOF
+		return io.EOF
 	}
-	var (
-		block  DerivationBlock
-		idx    int
-		blocks []BlockAttributes
-	)
-	firstL2BlockNumber := big.NewInt(0).SetUint64(b.pendingBlocks[0].BlockNumber())
-	numBytes := uint64(rlp.IntSize(firstL2BlockNumber.Uint64()))
-
-	// Iterate block-by-block to enforce soft cap on batch size.
-	for idx, block = range b.pendingBlocks {
-		blockTimestamp := big.NewInt(0).SetUint64(block.Timestamp())
-		// Calculate size of txData.
-		txListSize := uint64(0)
-		for _, tx := range block.Txs() {
-			txListSize += uint64(len(tx))
-		}
-		blocks = append(blocks, BlockAttributes{blockTimestamp, block.Txs()})
-		numBytes += rlp.ListSize(uint64(rlp.IntSize(blockTimestamp.Uint64())) + rlp.ListSize(txListSize))
-
-		// Enforce soft cap on batch size.
-		if rlp.ListSize(numBytes) > b.maxBatchSize {
-			log.Info("Reached max batch size", "numBytes", numBytes, "maxBatchSize", b.maxBatchSize, "numBlocks", len(blocks))
-			break
+	// Process pending blocks.
+	numProcessed := 0
+	for _, block := range b.pendingBlocks {
+		if err := b.encoder.ProcessBlock(&block); err != nil {
+			if errors.Is(err, errors.New("full batch")) {
+				break
+			}
 		}
+		numProcessed += 1
 	}
-	// Construct batch attributes.
-	attrs := &BatchAttributes{firstL2BlockNumber, blocks}
-	log.Info("Serialized l2 blocks", "first", firstL2BlockNumber, "last", block.BlockNumber())
+	log.Info("Encoded l2 blocks", "num_processed", numProcessed)
 	// Advance queue.
-	b.pendingBlocks = b.pendingBlocks[idx+1:]
+	b.pendingBlocks = b.pendingBlocks[numProcessed:]
 	log.Trace("Advanced pending blocks", "len", len(b.pendingBlocks))
-	return attrs, nil
+	return nil
+}
+
+type DecodeTxBatchError struct{ msg string }
+
+func (e *DecodeTxBatchError) Error() string {
+	return fmt.Sprintf("failed to decode batch: %s", e.msg)
 }
 
-func encodeBatch(b *BatchAttributes) ([]byte, error) {
-	var w bytes.Buffer
-	// Batch starts with version byte
-	if err := w.WriteByte(TxBatchVersion()); err != nil {
-		return nil, err
+// TODO: this is not currently called anywhere but will be useful for testing.
+func DecodeBatch(data []byte) (interface{}, error) {
+	if len(data) == 0 {
+		return nil, &DecodeTxBatchError{"empty batch data"}
+	}
+	version := data[0]
+	switch version {
+	case 0:
```

Comment:
use the `v0` definition. also I would like to use all upper-case identifiers to mark it as a constant, like `V0` or `BATCH_V0`. separately, its better to use a table to dispatch the decode instead of `switch` statement, but since we only have 1 version and its for testing purpose to decode its fine now